### DISCUSSION TITLE
Caching strategies for CKAN sites

### DISCUSSION THREAD

**Discussion by amercader** [ckan/ckan#6445] [Q&A]
Do you use a caching proxy like Squid, Nginx etc? A CDN? An external service?

What do you cache and what you don't? (eg logged in users? searches? dataset pages?)
What policies do you use? eg expiration times, triggers...

Would love to hear any experiences on this front

**Comment by ccancellieri** [comment]
Ciao @amercader so after some attempt I'm able to reply on this even if I'm not convinced this is the perfect solution.
Even thought it is giving me a great speedup and relatively low load on the ckan instance target

1.st I'm using nginx to proxy calls to ckan
2.nd Ckan 2.8.x is not supporting caching over blueprint (that I'm using in my plugin {CKAN}/terriajs/...
3.rd Ckan 2.8.x with cache enabled is working properly on all the rest of the endpoints (css, js, html etc are all properly cached on the client side)

So I decided to use your suggestion to have cache on server side (nginx) and improve cache headers for blueprint over 2.8.x

This is what I've:

```

#https://www.nginx.com/blog/nginx-caching-guide/
    location /{CKAN}/terriajs/ {

    # location ~* \.(?:css|js|json)$ {
        #Make sure your backend does not return Set-Cookie header. If Nginx sees it, it disables caching.
        #If this is your case, the best option is to fix your backend. When fixing the backend is not an option, it's possible to instru$
        proxy_ignore_headers "Set-Cookie";
    #    proxy_hide_header "Set-Cookie";
    
        proxy_ignore_headers cache-control;
        proxy_hide_header Cache-Control;
    
    #       proxy_cache_use_stale error timeout http_500 http_502 http_503 http_504;

            add_header Cache-Control: "public, max-age=60";
    #        add_header Cache-Control: "private";
    
        proxy_cache       my_cache;
    #    proxy_cache_key $request_uri;
        proxy_cache_key $scheme$proxy_host$request_uri$args;
    
        #I track HIT and MISS via
        add_header X-Cache-Status $upstream_cache_status;
    
        # this line instructs NGINX to cache any request where the application responds with a 200 OK status 
        proxy_cache_valid 200 120m;
        proxy_cache_min_uses 5;
    
        # Very important for microcache
        proxy_cache_background_update on;
        proxy_cache_revalidate on;
    
        # Very important for concurrency
        proxy_cache_lock on;
        proxy_cache_use_stale updating;

        proxy_pass http://127.0.0.1:8080/{CKAN}/terriajs/;
    }

```


You may note that most of these proxy directives may stay at server {} level and not at server { location {}}

Well I'm doing that since I'm behind a IAP proxy and my authentications and sessions renewals are provided by redirection.

Having that at server level was causing an infinite loop during redirection and to avoid that I may have implemented some exceptions for /user/login and logout or better control 302 responses at global level.

I'm not interested in that (especially on production) so I'm using the above to limit cache usage to my custom endpoints /{CKAN}/terriajs/...

**Reply by ccancellieri** [reply]
I would suggest (for a stable deployments) to use etags for static assets and expiration time (not too long) for generated assets so it will be easy to update and regenerate new contents.

**Comment by Zharktas** [comment]
We at https://www.opendata.fi/en use cloudfront CDN in front of the service itself. We have configured cache headers of ckan to announce all anonymous requests to be public so that any page will be cached in cloudfront for awhile after the request. The caching time is not long to accommodate updates to datasets, as invalidating cloudfront cache is not simple thing to do. Any signed in user have private cache headers in use so none those pages will be cached in cloudfront. We no longer cache on the nginx level at all, as most of the requests never arrive on virtual machine itself.

As we run ckan 2.8 still for awhile, we have patched flask cache control headers ourselves to our instances as it never was backported to 2.8.
